File: test/test_api_integration.py
"""Tests for the external API integration foundation."""

from __future__ import annotations

import io
import unittest
from urllib.error import HTTPError, URLError

from model.api import APIConfig, APIRequest, APIResponse
from utils.api_client import ExternalAPIError, build_api_url, send_json_request


class FakeHTTPResponse:
    """Minimal urllib-like response for client tests."""

    def __init__(
        self,
        *,
        body: bytes,
        status: int = 200,
        headers: dict[str, str] | None = None,
    ) -> None:
        self._body = body
        self.status = status
        self.headers = headers or {}

    def __enter__(self) -> "FakeHTTPResponse":
        return self

    def __exit__(self, exc_type: object, exc: object, traceback: object) -> None:
        return None

    def read(self) -> bytes:
        return self._body


class APIModelTest(unittest.TestCase):
    """Validate transport-neutral API models."""

    def test_config_merges_default_auth_and_request_headers(self) -> None:
        config = APIConfig(
            base_url="https://api.example.com/",
            default_headers={"X-App": "dashboard"},
            api_key="secret-token",
        )

        headers = config.headers({"X-Request": "lead-sync"})

        self.assertEqual(config.normalized_base_url, "https://api.example.com")
        self.assertEqual(headers["Accept"], "application/json")
        self.assertEqual(headers["X-App"], "dashboard")
        self.assertEqual(headers["Authorization"], "Bearer secret-token")
        self.assertEqual(headers["X-Request"], "lead-sync")

    def test_request_normalizes_method(self) -> None:
        request = APIRequest(method="post", path="/leads", json={"name": "Ada"})

        self.assertEqual(request.method, "POST")

    def test_response_ok_flags_success_statuses(self) -> None:
        self.assertTrue(APIResponse(status_code=204).ok)
        self.assertFalse(APIResponse(status_code=404).ok)


class APIClientTest(unittest.TestCase):
    """Validate reusable external API client helpers."""

    def test_build_api_url_includes_path_and_query(self) -> None:
        config = APIConfig(base_url="https://api.example.com/v1/")
        request = APIRequest(
            method="GET",
            path="consignments",
            query={"status": "open", "tag": ["a", "b"]},
        )

        self.assertEqual(
            build_api_url(config, request),
            "https://api.example.com/v1/consignments?status=open&tag=a&tag=b",
        )

    def test_send_json_request_posts_body_and_headers(self) -> None:
        captured: dict[str, object] = {}

        def opener(request: object, *, timeout: float) -> FakeHTTPResponse:
            captured["url"] = request.full_url
            captured["method"] = request.get_method()
            captured["body"] = request.data
            captured["content_type"] = request.headers["Content-type"]
            captured["timeout"] = timeout
            return FakeHTTPResponse(
                body=b'{"id":123}',
                status=201,
                headers={"X-Trace": "abc"},
            )

        config = APIConfig(base_url="https://api.example.com", timeout_seconds=3)
        request = APIRequest(method="POST", path="/leads", json={"name": "Ada"})

        response = send_json_request(config, request, opener=opener)

        self.assertEqual(captured["url"], "https://api.example.com/leads")
        self.assertEqual(captured["method"], "POST")
        self.assertEqual(captured["body"], b'{"name":"Ada"}')
        self.assertEqual(captured["content_type"], "application/json")
        self.assertEqual(captured["timeout"], 3)
        self.assertEqual(response.status_code, 201)
        self.assertEqual(response.data, {"id": 123})
        self.assertEqual(response.headers["X-Trace"], "abc")

    def test_send_json_request_raises_external_api_error_for_http_error(self) -> None:
        def opener(request: object, *, timeout: float) -> FakeHTTPResponse:
            raise HTTPError(
                request.full_url,
                400,
                "Bad Request",
                hdrs={},
                fp=io.BytesIO(b'{"message":"invalid"}'),
            )

        with self.assertRaises(ExternalAPIError) as ctx:
            send_json_request(
                APIConfig(base_url="https://api.example.com"),
                APIRequest(method="GET", path="/broken"),
                opener=opener,
            )

        self.assertEqual(ctx.exception.status_code, 400)
        self.assertEqual(ctx.exception.details["response"], {"message": "invalid"})

    def test_send_json_request_raises_external_api_error_for_network_error(self) -> None:
        def opener(request: object, *, timeout: float) -> FakeHTTPResponse:
            raise URLError("timed out")

        with self.assertRaises(ExternalAPIError) as ctx:
            send_json_request(
                APIConfig(base_url="https://api.example.com"),
                APIRequest(method="GET", path="/slow"),
                opener=opener,
            )

        self.assertEqual(ctx.exception.status_code, 502)
        self.assertEqual(ctx.exception.details["reason"], "timed out")


if __name__ == "__main__":
    unittest.main()

File: test/test_api_integration.js
import assert from 'node:assert/strict';
import test from 'node:test';

import { APIConfig, APIRequest, APIResponse } from '../model/api.js';
import { ExternalAPIError, buildApiUrl, sendJsonRequest } from '../utils/apiClient.js';

test('APIConfig merges default, auth, and request headers', () => {
  const config = new APIConfig({ baseUrl: 'https://api.example.com/', defaultHeaders: { 'X-App': 'dashboard' }, apiKey: 'secret-token' });
  const headers = config.headers({ 'X-Request': 'lead-sync' });
  assert.equal(config.normalizedBaseUrl, 'https://api.example.com');
  assert.equal(headers.Accept, 'application/json');
  assert.equal(headers['X-App'], 'dashboard');
  assert.equal(headers.Authorization, 'Bearer secret-token');
  assert.equal(headers['X-Request'], 'lead-sync');
});

test('APIRequest normalizes method', () => {
  assert.equal(new APIRequest({ method: 'post', path: '/leads', json: { name: 'Ada' } }).method, 'POST');
});

test('APIResponse ok flags success status codes', () => {
  assert.equal(new APIResponse({ statusCode: 204 }).ok, true);
  assert.equal(new APIResponse({ statusCode: 404 }).ok, false);
});

test('buildApiUrl includes path and repeated query values', () => {
  const config = new APIConfig({ baseUrl: 'https://api.example.com/v1/' });
  const request = new APIRequest({ method: 'GET', path: 'consignments', query: { status: 'open', tag: ['a', 'b'] } });
  assert.equal(buildApiUrl(config, request), 'https://api.example.com/v1/consignments?status=open&tag=a&tag=b');
});

test('sendJsonRequest posts JSON body and headers', async () => {
  const captured = {};
  const transport = async (url, init) => {
    captured.url = url;
    captured.method = init.method;
    captured.body = init.body;
    captured.contentType = init.headers['Content-Type'];
    return new Response('{"id":123}', { status: 201, headers: { 'X-Trace': 'abc' } });
  };
  const response = await sendJsonRequest(new APIConfig({ baseUrl: 'https://api.example.com', timeoutSeconds: 3 }), new APIRequest({ method: 'POST', path: '/leads', json: { name: 'Ada' } }), { transport });
  assert.equal(captured.url, 'https://api.example.com/leads');
  assert.equal(captured.method, 'POST');
  assert.equal(captured.body, '{"name":"Ada"}');
  assert.equal(captured.contentType, 'application/json');
  assert.equal(response.statusCode, 201);
  assert.deepEqual(response.data, { id: 123 });
  assert.equal(response.headers['x-trace'], 'abc');
});

test('sendJsonRequest raises ExternalAPIError for HTTP errors', async () => {
  const transport = async () => new Response('{"message":"invalid"}', { status: 400 });
  await assert.rejects(
    () => sendJsonRequest(new APIConfig({ baseUrl: 'https://api.example.com' }), new APIRequest({ method: 'GET', path: '/broken' }), { transport }),
    (error) => error instanceof ExternalAPIError && error.statusCode === 400 && error.details.response.message === 'invalid',
  );
});

test('sendJsonRequest raises ExternalAPIError for network errors', async () => {
  const transport = async () => { throw new Error('timed out'); };
  await assert.rejects(
    () => sendJsonRequest(new APIConfig({ baseUrl: 'https://api.example.com' }), new APIRequest({ method: 'GET', path: '/slow' }), { transport }),
    (error) => error instanceof ExternalAPIError && error.statusCode === 502 && error.details.reason === 'timed out',
  );
});
